File: pr049.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
The arithmetic sequence, 1487, 4817, 8147, in which each of the terms increases
by 3330, is unusual in two ways: (i) each of the three terms are prime, and,
(ii) each of the 4-digit numbers are permutations of one another.

There are no arithmetic sequences made up of three 1-, 2-, or 3-digit primes,
exhibiting this property, but there is one other 4-digit increasing sequence.

What 12-digit number do you form by concatenating the three terms in this
sequence?
"""

from prime import primes


def is4digit(num):
  """num is 4-digit number or not"""
  return num >= 1000 and num < 10000


def permutations(n1, n2, n3):
  """check three number is permutations or not"""
  l1 = list(str(n1))
  l1.sort()
  l2 = list(str(n2))
  l2.sort()
  l3 = list(str(n3))
  l3.sort()
  return "".join(l1) == "".join(l2) == "".join(l3)


def found_3arithmetic(lst):
  """find 3 arithmetic sequence from lst"""
  arithmetics = []
  length = len(lst)
  for i in range(length):
    num1 = lst[i]
    for j in range(i+1, length):
      num2 = lst[j]
      num3 = num2 + (num2 - num1)
      if lst.count(num3) > 0:
        if permutations(num1, num2, num3):
          arithmetics.append([num1, num2, num3])
  return arithmetics


def main():
  """main function"""
  primes_4digits = filter(is4digit, primes(10000))
  ars = found_3arithmetic(primes_4digits)
  # found unknonw triples.
  for ar in ars:
    if ar[0] != 1487:
      print reduce(lambda a,b:a+b, map(str, ar))
      break;  # out to loop


if __name__ == '__main__':
  main()

File: pr049.hs
{-
The arithmetic sequence, 1487, 4817, 8147, in which each of the terms increases
by 3330, is unusual in two ways: (i) each of the three terms are prime, and,
(ii) each of the 4-digit numbers are permutations of one another.

There are no arithmetic sequences made up of three 1-, 2-, or 3-digit primes,
exhibiting this property, but there is one other 4-digit increasing sequence.

What 12-digit number do you form by concatenating the three terms in this
sequence?
-}

module Main where

import Control.Applicative   -- for <$>, <*>
import Data.List  -- for sort
import Prime (primesUnder)

-- 4-digits primes
primes4digits :: [Int]
primes4digits = dropWhile (<999) $ primesUnder 10000

-- 3 element list is arithmetic ?
arithmetic :: [Int] -> Bool
arithmetic (n1:n2:n3:_) = (n2 > n1) && ((n2 - n1) == (n3 - n2))

-- two number is permutations ?
isPermutation :: Int -> Int -> Bool
isPermutation n m = (sort . show) n == (sort . show) m

-- XXX: 3 numbers from lst
-- XXX:   I try to use an applicative style, but it is slow...
__threes :: [Int] -> [[Int]]
__threes ns = map (\(n1,n2,n3) -> [n1,n2,n3]) triples
            where
              triples = (,,) <$> ns <*> ns <*> ns

-- choice three number from list. the three number is permutation
threes :: [Int] -> [[Int]]
threes ns = filterThree $ foldr1 (++) $ map sub [0..(length ns - 3)]
            where
              filterThree = filter (\ns -> length ns == 3)
              sub x = map ((ns !! x):) (choicePair (filter (isPermutation (ns !! x)) (drop (x+1) ns)))

-- choice two number from list
choicePair :: [Int] -> [[Int]]
choicePair ns = map sort $ concatMap func [0..(length ns - 1)]
                where
                  func x = map (:[ns !! x]) $ drop (x+1) ns

-- concat numbers
concatNums :: [Int] -> Int
concatNums ns = read $ concatMap show ns

-- main function
main :: IO ()
main = print $ concatNums targetTriple
       where
         arithmeticSequence = filter arithmetic $ threes primes4digits
         targetTriple = head $ filter (\t -> head t /= 1487) arithmeticSequence
